README.md
# Ansible Versioning Test Collection

This is for testing collection versioning history with changelogs.

## Tested with Ansible

Tested with both the current development version of Ansible.

## External requirements

No requirements.

## Included content

- ``foo`` module
- ``bar`` module
- ``baz`` module
- ``echo`` module
- ``reverse`` lookup
- ``bob`` lookup

## Using this collection

See [Ansible Using collections](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html) for more details.

## Release notes

See [here](https://github.com/felixfontein/ansible-versioning_test_collection/tree/main/CHANGELOG.rst).

## More information

- [Ansible Collection overview](https://github.com/ansible-collections/overview)
- [Ansible User guide](https://docs.ansible.com/ansible/latest/user_guide/index.html)
- [Ansible Developer guide](https://docs.ansible.com/ansible/latest/dev_guide/index.html)
- [Ansible Community code of conduct](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)

## Licensing

GNU General Public License v3.0 or later.

See [COPYING](https://www.gnu.org/licenses/gpl-3.0.txt) to see the full text.


CHANGELOG.rst
==================================================================
Felix Fontein's Amazing Versioning Test Collection 2 Release Notes
==================================================================

.. contents:: Topics

This changelog describes changes after version 1.0.2.

v2.3.0-pre
==========

Release Summary
---------------

Just testing with pre-releases.

v2.2.0
======

Release Summary
---------------

Testing long changelog fragments. Nothing else in this release. Ignore the fragment contents!

Minor Changes
-------------

- And a final complex reStructuredText fragment.

  The way to import a PowerShell or C# module util from a collection has changed in the Ansible 2.9 release. In Ansible
  2.8 a util was imported with the following syntax:

  .. code-block:: powershell

      #AnsibleRequires -CSharpUtil AnsibleCollections.namespace_name.collection_name.util_filename
      #AnsibleRequires -PowerShell AnsibleCollections.namespace_name.collection_name.util_filename

  In Ansible 2.9 this was changed to:

  .. code-block:: powershell

      #AnsibleRequires -CSharpUtil ansible_collections.namespace_name.collection_name.plugins.module_utils.util_filename
      #AnsibleRequires -PowerShell ansible_collections.namespace_name.collection_name.plugins.module_utils.util_filename

  The change in the collection import name also requires any C# util namespaces to be updated with the newer name
  format. This is more verbose but is designed to make sure we avoid plugin name conflicts across separate plugin types
  and to standardise how imports work in PowerShell with how Python modules work.
- This is an example of a longer reStructuredText fragment.

  Ansible 2.9 handles "unsafe" data more robustly, ensuring that data marked "unsafe" is not templated. In previous versions, Ansible recursively marked all data returned by the direct use of ``lookup()`` as "unsafe", but only marked structured data returned by indirect lookups using ``with_X`` style loops as "unsafe" if the returned elements were strings. Ansible 2.9 treats these two approaches consistently.

  As a result, if you use ``with_dict`` to return keys with templatable values, your templates may no longer work as expected in Ansible 2.9.

  To allow the old behavior, switch from using ``with_X`` to using ``loop`` with a filter as described at :ref:`migrating_to_loop`.
- This is an example of a more complex reStructuredText fragment.

  Module and module_utils files can now use relative imports to include other module_utils files.
  This is useful for shortening long import lines, especially in collections.

  Example of using a relative import in collections:

  .. code-block:: python

    # File: ansible_collections/my_namespace/my_collection/plugins/modules/my_module.py
    # Old way to use an absolute import to import module_utils from the collection:
    from ansible_collections.my_namespace.my_collection.plugins.module_utils import my_util
    # New way using a relative import:
    from ..module_utils import my_util

  Modules and module_utils shipped with Ansible can use relative imports as well but the savings
  are smaller:

  .. code-block:: python

    # File: ansible/modules/system/ping.py
    # Old way to use an absolute import to import module_utils from core:
    from ansible.module_utils.basic import AnsibleModule
    # New way using a relative import:
    from ...module_utils.basic import AnsibleModule

  Each single dot (``.``) represents one level of the tree (equivalent to ``../`` in filesystem relative links).

  .. seealso:: `The Python Relative Import Docs <https://www.python.org/dev/peps/pep-0328/#guido-s-decision>`_ go into more detail of how to write relative imports.

Bugfixes
--------

- Renamed ``master`` git branch to ``main``.

v2.1.0
======

Release Summary
---------------

Bob was there, too!

Bugfixes
--------

- bob lookup - forgot to check whether ``Bob`` was already there.

New Plugins
-----------

Lookup
~~~~~~

- bob - Bob was there, too

v2.0.0
======

Release Summary
---------------

We're happy to release 2.0.0 with a new plugin!

Bugfixes
--------

- reverse lookup - fix bug in error message.

New Plugins
-----------

Lookup
~~~~~~

- reverse - reverse magic


plugins/lookup/bob.py
from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = '''
lookup: bob
author: Felix Fontein
version_added: '2.1.0'
short_description: Bob was there, too
description:
  - Adds Bob.
options:
  _terms:
    description: Strings to play with.
    type: list
    elements: str
    required: True
'''

EXAMPLES = """
- debug: msg="{{ lookup('felixfontein.versioning_test_collection.bob', 'this', 'is', 'a', 'test') }}"
"""

RETURN = """
  _raw:
    description:
      - List with Bob.
    type: list
    elements: str
"""

from ansible.errors import AnsibleError
from ansible.plugins.lookup import LookupBase


class LookupModule(LookupBase):

    def run(self, terms, variables, **kwargs):
        if not isinstance(terms, list):
            raise AnsibleError("reverse expects a list")

        if 'Bob' in terms:
            return terms

        return terms + ['Bob']


plugins/modules/bar.py
#!/usr/bin/python
#
# Copyright 2020 Felix Fontein
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}


DOCUMENTATION = r'''
---
module: bar
short_description: Make a GET
version_added: '0.2.0'
description:
  - Makes a GET to a URL.
  - This is really only a test.
author:
  - "Felix Fontein (@felixfontein)"
requirements:
  - An internet connection
options:
  url:
    description:
      - The URL to POST to.
    required: yes
    type: str
'''

EXAMPLES = r'''
- name: Do something
  felixfontein.versioning_test_collection.bar:
    url: https://example.com
  register: result
- debug:
    msg: |
      {{ result.status }} --
      {{ result.body }}
'''

RETURN = r'''
status:
  description:
    - The status code of the POST request.
  returned: always
  type: int
  sample: 200
body:
  description:
    - The parsed JSON result.
  returned: always
  type: dict
  sample: '{"key": "value"}'
'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url


def main():
    argument_spec = dict(
        url=dict(type='str', required=True),
    )
    module = AnsibleModule(argument_spec)

    r, i = fetch_url(
        module,
        module.params['url'],
        method='GET',
    )

    status = i["status"]
    try:
        body = r.read()
    except Exception as dummy:
        body = i.get('body')

    if i.get('content-type') != 'application/json':
        module.fail_json(status=status, msg='Expected JSON return value, but got type "{0}"'.format(i.get('content-type')))

    try:
        data = module.from_json(body)
    except Exception as e:
        module.fail_json(status=status, msg='Error while decoding JSON: {0}'.format(e))

    module.exit_json(
        status=status,
        body=data
    )


if __name__ == '__main__':
    main()


plugins/modules/baz.py
#!/usr/bin/python
#
# Copyright 2020 Felix Fontein
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}


DOCUMENTATION = r'''
---
module: baz
short_description: Make a HEAD
version_added: '1.0.0'
description:
  - Makes a HEAD to a URL.
  - This is really only a test.
author:
  - "Felix Fontein (@felixfontein)"
requirements:
  - An internet connection
options:
  url:
    description:
      - The URL to POST to.
    required: yes
    type: str
'''

EXAMPLES = r'''
- name: Do something
  felixfontein.versioning_test_collection.baz:
    url: https://example.com
  register: result
- debug:
    msg: '{{ result.status }}'
'''

RETURN = r'''
status:
  description:
    - The status code of the HEAD request.
  returned: always
  type: int
  sample: 200
'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url


def main():
    argument_spec = dict(
        url=dict(type='str', required=True),
    )
    module = AnsibleModule(argument_spec)

    r, i = fetch_url(
        module,
        module.params['url'],
        method='HEAD',
    )

    module.exit_json(
        status=i["status"],
    )


if __name__ == '__main__':
    main()


plugins/modules/foo.py
#!/usr/bin/python
#
# Copyright 2020 Felix Fontein
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}


DOCUMENTATION = r'''
---
module: foo
short_description: Make a POST
description:
  - Makes a POST to a URL.
  - This is really only a test.
author:
  - "Felix Fontein (@felixfontein)"
requirements:
  - An internet connection
options:
  url:
    description:
      - The URL to POST to.
    required: yes
    type: str
  data:
    description:
      - Data to send in the POST request.
    type: str
    default: "x=42"
    version_added: '0.2.0'
'''

EXAMPLES = r'''
- name: Do something
  felixfontein.versioning_test_collection.foo:
    url: https://example.com
  register: result
- debug:
    msg: |
      {{ result.status }} --
      {{ result.body }}
'''

RETURN = r'''
status:
  description:
    - The status code of the POST request.
  returned: always
  type: int
  sample: 200
body:
  description:
    - The parsed JSON result.
  returned: always
  type: dict
  sample: '{"key": "value"}'
'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url


def main():
    argument_spec = dict(
        url=dict(type='str', required=True),
        data=dict(type='str', default='x=42'),
    )
    module = AnsibleModule(argument_spec)

    r, i = fetch_url(
        module,
        module.params['url'],
        method='POST',
        data=module.params['data'],
    )

    status = i["status"]
    try:
        body = r.read()
    except Exception as dummy:
        body = i.get('body')

    if i.get('content-type') != 'application/json':
        module.fail_json(status=status, msg='Expected JSON return value, but got type "{0}"'.format(i.get('content-type')))

    try:
        data = module.from_json(body)
    except Exception as e:
        module.fail_json(status=status, msg='Error while decoding JSON: {0}'.format(e))

    module.exit_json(
        status=status,
        body=data
    )


if __name__ == '__main__':
    main()


plugins/lookup/reverse.py
from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = '''
lookup: reverse
author: Felix Fontein
version_added: '2.0.0'
short_description: reverse magic
description:
  - Given a list of strings, will do some reverse magic.
options:
  _terms:
    description: Strings to play with.
    type: list
    elements: str
    required: True
'''

EXAMPLES = """
- debug: msg="{{ lookup('felixfontein.versioning_test_collection.reverse', 'this', 'is', 'a', 'test') }}"
"""

RETURN = """
  _raw:
    description:
      - Reverse-magic'ed list.
    type: list
    elements: str
"""

from ansible.errors import AnsibleError
from ansible.plugins.lookup import LookupBase


class LookupModule(LookupBase):

    def run(self, terms, variables, **kwargs):
        if not isinstance(terms, list):
            raise AnsibleError("reverse expects a list")

        return [str(f)[::-1] for f in reversed(terms)]


plugins/modules/echo.py
#!/usr/bin/python
#
# Copyright 2020 Felix Fontein
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}


DOCUMENTATION = r'''
---
module: echo
short_description: Echo params
version_added: '1.0.0'
description:
  - Echo module params.
author:
  - "Felix Fontein (@felixfontein)"
options:
  params:
    description:
      - Some params.
    required: yes
    type: dict
'''

EXAMPLES = r'''
- name: Do something
  felixfontein.versioning_test_collection.echo:
    params:
      foo: bar
      weee: true
      nothing: null
  register: result
- debug:
    msg: 'Echo! {{ result.status }}'
'''

RETURN = r'''
params:
  description:
    - What was passed.
  returned: always
  type: dict
  sample: {}
'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url


def main():
    argument_spec = dict(params=dict(type='dict', required=True))
    module = AnsibleModule(argument_spec)
    module.exit_json(params=module.params["params"])


if __name__ == '__main__':
    main()
